### DISCUSSION TITLE
Server -> Client Invocations

### DISCUSSION THREAD

**Discussion by IAmUser4574** [icerpc/icerpc-csharp#3960] [Q&A] [answered]
With the existing API documentation and examples I was able to write a few successful test applications of Client -> Server invocations and responses. However, I'd like to support Server-originated messages to Clients and Server forwarding of messages from one client to another. I've poured through the docs and, while I've found references to a `ConnectionCache` (which I imagine I can implement on the Server), I've yet to get something like that working with client callbacks as the documentation is sparse on usage, both on the setup of the server and the client.

I was wondering if anyone familiar with this project has already tread down this path and figured out a working configuration with client callbacks.

**Comment by pepone** [comment, marked as answer]
You should read https://zeroc.com/blog/other-way-around and check the companion example https://github.com/icerpc/icerpc-csharp/tree/main/examples/slice/Thermostat, I think that is what you are looking for!

Let us know if something is not clear.

**Comment by bernardnormier** [comment]
`ConnectionCache` is a class provided by IceRPC - you don't implement it. And it's a client-side class: it's a cache of outbound connections.

See
https://docs.icerpc.dev/icerpc/connection/how-to-create-a-connection#creating-a-client-connection
and
https://docs.icerpc.dev/api/csharp/api/IceRpc.ConnectionCache.html
